Allow to create custom factories for structs

Artpej opened the pull request:
With this, the Factories generated by Golo can be overloaded by custom ones.

for example:

```golo
module test

struct Point = { x, y }

function Point = -> test.types.Point(0,0)

function main = |args| {
  println(Point(1, 2))
  println(Point())
}

```
running it prints the following console output:

```
struct Point{x=1, y=2}
struct Point{x=0, y=0}
```

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ir/GoloFunction.java:
```diff
@@ -334,4 +335,30 @@ public void walk(GoloIrVisitor visitor) {
+    hash = 89 * hash + Objects.hashCode(this.getArity());
+    return hash;
+  }
+
```
why don't you just use `Objects.hash(name, getArity())`?

Artpej replied:
No idea, should I be less confident with my IDE that do `hashCode` automatically like this?

yloiseau replied:
I think it's functionally equivalent, it just seems more readable to me.

jponge replied:
IDE do it right usually, especially when you specify which fields are nullable :+1:

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ir/GoloFunction.java:
```diff
@@ -334,4 +335,30 @@ public void walk(GoloIrVisitor visitor) {
+      return false;
+    }
+    return true;
+  }
```
`return Objects.equals(this.name, other.name) && this.getArity() == other.getArity()` ? :smile:

Artpej replied:
We have to check that `other` is not `null` and that it is a `GoloFunction` instance :wink:

yloiseau replied:
Yes but I mean after the cast, you can just return the boolean expression instead of the two `if`s (since `&&` is lazy)

jponge replied:
Same here, IDEs do it right.

But Java 8 `Object.equals` is also fine.

yloiseau commented on src/main/java/org/eclipse/golo/compiler/SugarExpansionVisitor.java:
```diff
@@ -315,4 +315,14 @@ public void visitDestructuringAssignment(DestructuringAssignment assignment) {
+  public void visitStruct(Struct struct) {
+    module.addFunctions(struct.createFactories());
+  }
+
```
My previous refactoring helped, didn't it? :wink:

Artpej replied:
indeed! :wink:

yloiseau commented:
I think more criteria should be used to `equals`/`hashCode` functions, e.g. at least scope, maybe visibility.

For instance:
```golo
module Plop 

struct Foo = { x, y }
augment Foo {
  function plop = |this| -> "ploping: " + this: x() + ", " + this: y()
}

function plop = |val| -> "Plop with " + val
```
the two `plop` functions would be considered equals. Not an issue for your change, but I think it can latter leads to strange behavior.

danielpetisme commented:
I <3 this PR, 
is it out of the box compatible with Immutable structs ?

yloiseau commented:
It works by *not* injecting the factory methods in the module if an equal one (same name and arity) is present, so yes, if you define an `ImmutableMyStruct` with the right arity, it will “override” the default one.

yloiseau commented:
This will indeed allow to check preconditions on instanciation :smile:.
Again about not taking account of visibility in the `equals`, it can be a feature (or not). Indeed, if you do:
```golo
module Test

struct MyStruct = {x, y}

local function MyStruct = |x, y| -> Test.types.MyStruct(x, y)
local function MyStruct -> Test.types.MyStruct(0, 0)
```
the local function will be considered equal to default factories, and thus they will not be injected. That is, you can't create a `MyStruct` like:
```golo
module Use

import Test

function main = |args| {
  let s = MyStruct(1, 2) # will fail
  let t = Test.types.MyStruct(1, 2) # should be ok
}
```
It can be seen as a feature :wink:

Artpej commented:
@danielpetisme there's 2 ways to "override" the immutables factories:

#### With a frozen copy
```golo
module test

struct Point = { x, y }

function Point = -> test.types.Point(0,0)

function ImmutablePoint = |a,b| -> test.types.Point(a,b): frozenCopy()

function main = |args| {
  var point = Point(1, 2)
  println(point + " " + point: isFrozen())
  point = ImmutablePoint(1,2)
  println(point + " " + point: isFrozen())
}
```

#### Or with internals knowleges :wink:
```golo
module test

struct Point = { x, y }

function Point = -> test.types.Point(0,0)

function ImmutablePoint = |a,b| -> test.types.Point.$_immutable(a,b)

function main = |args| {
  var point = Point(1, 2)
  println(point + " " + point: isFrozen())
  point = ImmutablePoint(1,2)
  println(point + " " + point: isFrozen())
}

```

@yloiseau i think function name + arity is suffisent for function discrimination inside a module. The unwanted behaviour is that if we declare tow times a function with same name and arity only the first declared one will be used without warn at compilation time. (before this PR an exception was thrown at runtime: `Exception in thread "main" java.lang.ClassFormatError: Duplicate method name&signature in class file`)

Artpej commented:
@yloiseau ok I just had a look on #312 comments and we should take in account variable arity too :wink:

Maybe we could warn at compilation when a function is suceptible to hide another one (except for the struct factories since it's a feature :smile: )

yloiseau commented:
I think that at least a warning is needed. I'd argue it should be a *compile time* error. Checking can be as simple as:
```java
public void addFunction(GoloFunction function) {
    if (functions.contains(function) && !function.isSynthetic()){
        // throw ....
    }
    ...
  }
```

since generated default factories are synthetic, they are just ignored, but a user defined function would failed.

yloiseau commented:
see https://gist.github.com/yloiseau/464e283d85e2ae3d91dd, I created that to test some thing while refactoring IR (don't remember what), but did'nt used it. 
Maybe instead of overriding `GoloFunction.equals`, the module functions can be a `Map<FunctionSignature, GoloFunction>`

jponge commented:
Very good pull-request!

jponge commented:
I concur with @yloiseau that function hiding shall raise a compilation error.

jponge commented:
Assigned to @Artpej, please ping us when a new review round is possible :smile:

yloiseau commented:
Let's continue the discussion on ambiguous function references on the golo-dev ML since the scope is larger than this PR...

jponge commented:
Ping, any news?

Artpej commented:
Still working one this :
 - Add compilation error when a function hide another one (fixed arity)
 - keep the same function resoluion order as Java for varargs ([reference](http://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12.2.4))

I will try to push an update for this proposal by the end of the week.

jponge commented:
:+1:

Artpej commented:
I just updated this one: 

- [x] Add compilation error when a function hide another one (fixed arity)
- [x] fixed arity function are now resolved before varargs one 

I think about a huge refactoring for functions resolution based on works that are logged here  #326  thanks to @yloiseau :wink: but it's more complexe work than this one...

jponge commented:
Thanks @Artpej 

Are we good to go on this one?

Artpej commented:
Seem's to me fully fonctionnal like this. It is not as elegant as I expected, but a refactoring is coming. 
Ready for review.

danielpetisme commented:
I'm really impatient to use this feature for the testing module, it will really help!
I was bored to have a "init" method to override the no-args constructor. 
Thanks @Artpej

jponge commented on doc/structs.asciidoc:
```diff
@@ -102,6 +102,57 @@ function main = |args| {
 
+=== Custom factories
+
+Golo generates two factories for the structures. One with member's initial values as parameter and
```
for structures

jponge commented on doc/structs.asciidoc:
```diff
@@ -102,6 +102,57 @@ function main = |args| {
+struct Point{x=null, y=null}
+----
+
+NOTE: By default the no argument factory sets every member to `null`
```
no-argument

jponge commented on src/main/java/org/eclipse/golo/compiler/ParseTreeToGoloIrVisitor.java:
```diff
@@ -120,7 +89,21 @@ public void leaveNamedAugmentation() {
 
     public void addFunction(GoloFunction function) {
-      this.functionContainersStack.peek().addFunction(function);
+      FunctionContainer container = this.functionContainersStack.peek();
```
:+1:

jponge commented:
Done my review with little doc edit suggestions

Artpej commented:
done! :beers:

yloiseau commented on doc/structs.asciidoc:
```diff
@@ -102,6 +102,57 @@ function main = |args| {
+
+struct Point = { x, y }
+
+function Point = -> test.types.Point(0,0)
```
Maybe you could add a note on how to call the original constructor (the `<module>.types.<struct>` thing) and how to override the immutable constructor (see @danielpetisme question in the PR comments).

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ParseTreeToGoloIrVisitor.java:
```diff
@@ -120,7 +89,21 @@ public void leaveNamedAugmentation() {
+        errorMessage(AMBIGUOUS_DECLARATION, function.getASTNode(),
+            String.format("Declaring a function `%s` twice (declared first at line=%d, column=%d) ",
+                function.getName(),
+                firstDeclaration.getASTNode().getLineInSourceCode(),
```
You can directly use `GoloASTNode::getPositionInSourceCode` which override `toString`...

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ParseTreeToGoloIrVisitor.java:
```diff
@@ -162,10 +145,36 @@ private LocalReference getOrCreateReference(ASTLetOrVar.Type type, String name,
+                              String message) {
+      String errorMessage = String.format(
+          "%s at (line=%d, column=%d)",
+          message, node.getLineInSourceCode(), node.getColumnInSourceCode());
```
same here, `GoloASTNode::getPositionInSourceCode`

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ParseTreeToGoloIrVisitor.java:
```diff
@@ -69,6 +37,7 @@ public Object visit(ASTerror node, Object data) {
     private Deque<FunctionContainer> functionContainersStack = new LinkedList<>();
     private Deque<Object> objectStack = new LinkedList<>();
     private Deque<ReferenceTable> referenceTableStack = new LinkedList<>();
+    private GoloCompilationException.Builder exceptionBuilder;
```
why not make it final and passed in the constructor (and remove the setter)?

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ir/GoloFunction.java:
```diff
@@ -344,4 +345,24 @@ public void walk(GoloIrVisitor visitor) {
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```
You don't test for `this == obj` ?

yloiseau commented on src/main/java/org/eclipse/golo/compiler/ir/GoloFunction.java:
```diff
@@ -344,4 +345,24 @@ public void walk(GoloIrVisitor visitor) {
+    if (!Objects.equals(this.name, other.name)) {
+      return false;
+    }
+    return this.getArity() == other.getArity();
```
You don't test for varargs ? So
```golo
function foo = |a| -> ...
function foo = |a...| -> ...
```

will fail to compile.

yloiseau commented on src/test/resources/for-execution/structs.golo:
```diff
@@ -80,3 +82,14 @@ function check_equals = ->
+
+function AList = -> golotest.execution.Structs.types.AList(ArrayList())
+
+function check_overload_factory = -> AList(): size()
```
You don't have tests for the new behaviors you added (compilation error if existant function and varargs ordering in method resolution)

Artpej commented:
Thanks @yloiseau ! (my bad, functions are still resolved in `FunctionCallSupport` not in a `RegularMathodFinder`)

Artpej commented:
@yloiseau , oups I forgot change the `toString`. I'll ping you when it is done.

jponge commented:
@yloiseau will have the final word on the code review :smile: 

Great work Sylvain and Yannick!

yloiseau commented:
:+1:

Artpej commented:
@yloiseau ping!

yloiseau commented:
ok for me

jponge commented:
It's in, thanks @Artpej @yloiseau

jponge commented:
Ooops, it broke the build, any clue @Artpej?

Artpej commented:
I'll have a look asap

yloiseau commented:
@Artpej I posted a patch on gitter…
